py/bitbox02/communication/u2fhid/u2fhid.py

```diff
@@ -0,0 +1,161 @@
+            raise Exception("Received error: invalid sequence")
+        if error_code == ERR_CHANNEL_BUSY:
+            raise Exception("Received error: channel busy")
+        if error_code == 0x7E:
+            raise Exception("Received error: encryption failed")
+        if error_code == ERR_OTHER:
+            raise Exception("Received unknown error")
+        raise Exception("Received error: %d" % error_code)
+
+    def write(self, data: bytes, endpoint: int, cid: int) -> None:
+        """
+        Send data to the device.
+
+        Args:
+            data: Data to send
+            endpoint: U2F HID command (endpoint selection)
+            cid: U2F HID channel ID
+        Throws:
+            ValueError: In case any value is out of range
+        """
+        if endpoint < 0 or endpoint > 0xFF:
+            raise ValueError("Channel command (endpoint) is out of range '0 < endpoint <= 0xFF'")
+        if cid < 0 or cid > 0xFFFFFFFF:
+            raise ValueError("Channel id is out of range '0 < cid <= 0xFFFFFFFF'")
+        data = bytearray(data)
+        data_len = len(data)
+        if data_len > 0xFFFF:
+            raise ValueError("Data is too large 'size <= 0xFFFF'")
+        seq = 0
+        idx = 0
+        buf = b""
+        # Allow to write an empty packet
+        single_empty_write = data_len == 0
+        while idx < data_len or single_empty_write:
+            if idx == 0:
+                # INIT frame
+                buf = data[idx : idx + min(data_len, USB_REPORT_SIZE - 7)]
+                self._device.write(
+                    b"\0"
+                    + struct.pack(">IBH", cid, endpoint, data_len)
+                    + buf
+                    + b"\xEE" * (USB_REPORT_SIZE - 7 - len(buf))
+                )
+            else:
+                # CONT frame
+                buf = data[idx : idx + min(data_len, USB_REPORT_SIZE - 5)]
+                self._device.write(
+                    b"\0"
+                    + struct.pack(">IB", cid, seq)
+                    + buf
+                    + b"\xEE" * (USB_REPORT_SIZE - 5 - len(buf))
+                )
+                seq += 1
+            idx += len(buf)
+            single_empty_write = False
+
+    def read(self, endpoint: int, cid: int) -> bytes:
+        """
+        Receive data from the device.
+
+        Args:
+            endpoint: The expected returned U2F HID command
+            cid: The expected returned U2F HID channel ID
+        Returns:
+            The read message combined from the u2fhid packets
+        Throws:
+            ValueError: In case any value is out of range
+            Exception: In case of USB communication issues
+        """
+        if endpoint < 0 or endpoint > 0xFF:
+            raise ValueError("Endpoint/U2F command is out of range '0 < endpoint <= 0xFF'")
+        if cid < 0 or cid > 0xFFFFFFFF:
+            raise ValueError("Channel id is out of range '0 < cid <= 0xFFFFFFFF'")
+        timeout_ms = 5000
```

Comment:
this timeout is too short, can't even unlock the device. the previous value was 5000000s (basically infinite). please restore it to this value.

Reply:
maybe hid device read can take an argument for no timeout (0, None, negative, ...).

Reply:
are there any other functional changes in the the non-usart code? I expected that the regular bitbox02/bootloader/usb/hid not to change at all, functionally.

Reply:
> this timeout is too short, can't even unlock the device. the previous value was 5000000s (basically infinite). please restore it to this value.

Was 5000s before - This is a typo due to missing the fact that the first argument was in seconds, and the `hid` argument is in ms instead. Will revert.

> are there any other functional changes in the the non-usart code? I expected that the regular bitbox02/bootloader/usb/hid not to change at all, functionally.

It should be unchanged, indeed. A quick diff (with whitespace ignored) shows no difference (apart from the 
timeout typo and changes needed for making the class compatible with the `TransportLayer` class).

Also there is the (pretty evident) rename of the generated files from `bitbox02/generated` to `communication/generated`.

Reply:
ok np ;) 

there was a hacky `timeout = timeout * 1000` before that dropped: https://github.com/digitalbitbox/bitbox02-firmware/pull/203/files#diff-420ffb3dcc6af5c0c0d4f7254829149dL146

it was seconds, not milliseconds, all along.

no one to blame but myself for quickly putting this in to fix the timeout issue right before release :D

Reply:
anyway, with 5s most of the blocking workflows, like reading the mnemonic, will fail.

Reply:
Fixed